### DISCUSSION TITLE
Use case of `--enclave-file` in running CCF host

### DISCUSSION THREAD

**Discussion by chintanr97** [microsoft/CCF#2704] [Q&A] [answered]
While running a new CCF host manually, using the `cchost` or `cchost.virtual` binaries, the `--enclave-file` option is required. Currently, I am little confused about its use case based on the explanation in command-line help. I would like to identify the difference between the "user application" which is deployed in the CCF network (for e.g., a sample JavaScript application) and the value passed to this argument.

IMO, if both of these imply the same property, i.e., the user application to be run on CCF, then how can we effectively build a library file for a JavaScript application using `cmake`? For example, I was attempting to use the `forum` application at `./samples/apps/forum` to be deployed by manually creating a node. For this, initially I tested the CCF node creation functionality using the command:

```
sudo /opt/ccf/bin/cchost.virtual  \
--enclave-file /opt/ccf/lib/libjs_generic.virtual.so \
--enclave-type virtual \
--consensus cft \
--rpc-address 127.0.0.1:<p1> \
--node-address 127.0.0.1:<p1> \
--ledger-dir ... \
--node-cert-file ... \
start \
--network-cert-file ... \
--member-info ... \
--constitution ...
```

The above command correctly started a CCF node with the specified enclave file or sample user application. However, I am missing out steps to run, let's say, the Forum JavaScript application on this node directly instead of `/opt/ccf/lib/libjs_generic.virtual.so`. 

For understanding a little deeper, I followed the `README.md` file [here](https://github.com/microsoft/CCF/blob/main/samples/apps/forum/README.md), to deploy the `Forum Application` as is. The app was deployed correctly with following logs in the shell:

```
 Starting 1 CCF node...
 Virtual mode enabled
 Started CCF network with the following nodes:
   Node [0] = https://127.0.0.1:8000
 You can now issue business transactions to the libjs_generic application
 Loaded JS application: dist
 Keys and certificates have been copied to the common folder: .workspace_ccf/sandbox_common
 See https://microsoft.github.io/CCF/main/use_apps/issue_commands.html for more information
 Press Ctrl+C to shutdown the network
```

I went over the `sandbox.sh` code, followed by `start_network.py` code in `/opt/ccf/bin` folder. But, I could not understand the difference between logging of `You can now issue business transactions to the libjs_generic application` and `Loaded JS application: dist`. Also, I could not locate a code snippet that would suggest that the JavaScript application was built as shared object and deployed to the network!

Please let me know if I have misunderstood something here, or if I am missing out something.

**Comment by achamayou** [comment, marked as answer]
`js_generic` is an enclave shared object that contains a reference javascript runtime. The idea is that all JavaScript applications use that, and upload their application code to the KV store, using a governance proposal (see https://microsoft.github.io/CCF/main/build_apps/js_app_bundle.html#deployment for details).

You may find the tutorial helpful (https://microsoft.github.io/CCF/main/build_apps/js_app_ts.html), but please do feel free to ask (or suggest improvements!) if anything is unclear.

**Reply by eddyashton** [reply]
To expand a little on what @achamayou said:

The `--enclave-file` argument specifies the shared object that the node should start try to load. In the case of a CCF written in C++, like the [logging sample app](https://github.com/microsoft/CCF/blob/main/samples/apps/logging/logging.cpp) or [TPCC testing app](https://github.com/microsoft/CCF/tree/main/src/apps/tpcc), this shared object completely defines a single application and the endpoints it exposes. If you start a service with `--enclave-file liblogging.enclave.so.signed`, then the code ID and SGX quote you get for each node on that service attest that it is running the code defined in `logging.cpp`, and so this quote covers what will happen if you call `POST /app/log/private`.

In the case of the `js_generic` app, the shared object that is loaded and covered by attestation is only the JS runtime, not the JS scripts that are run on top of that runtime. At the point where the infra prints `You can now issue business transactions to the libjs_generic application`, the service is fully functional but has no JS scripts loaded, and no application endpoints defined. These are added dynamically by additional governance transactions. After that point, the test infrastructure sends some additional transactions as the service operators which set the JS app (eg - a `set_js_app` proposal with the contents of the [forum sample app](https://github.com/microsoft/CCF/tree/main/samples/apps/forum)), and then prints `Loaded JS application` to let you know that those application endpoints are now available. But the attestation (the SGX quote and code ID) only covers the shared object that is initially loaded into SGX, which is the generic JS runtime, not the current state which may include some additional JS scripts. This makes the JS application far more flexible - you can update the JS scripts dynamically, adding endpoints or fixing bugs with governance proposals rather than node replacements - at the cost of making the audit story a little more complex. To know what happens when you call `POST /app/log/private`, you need to know the JS code that's currently running, not just the code ID. That's probably available in the ledger or from another endpoint defined by the JS app (but it could also be private, in the same way that the `.cpp`s used to build a C++ app _could_ be private, in which case your trust that this does something useful relies on your trust in the service operators).

Its also possible to bake a JS script directly into the enclave file that is built by CMake. Rather than adding governance proposals which modify script tables, and reading from those script tables for every incoming RPC, you could just load a single string that was defined at build time as the JS app. This would let you write your application code in JS, but get the attestation properties of the C++ model. This isn't something we have a sample of - we think the `js_generic` approach is more useful - but just wanted to call out that its absolutely possible if that's what you're looking to do.

**Reply by chintanr97** [reply]
I think this was really helpful. I was able to successfully deploy a new CCF node and create governance proposals to start the sample `Forum JS Application`. I will try on my end to achieve the same using CMake approach if required.

**Reply by iqsarv** [reply]
@achamayou @eddyashton It will be good to callout in the docs  that the proposal-based Deployment step should be done after the `js_generic` app is started using `cchost` for the app to be run in CCF network.

**Reply by ebo1981-cyber** [reply]
dear  eddyashton 
dear  achamayou
Is it possible for you to provide me with a video tutorial on installing and running the ccf environment  and application ?

**Reply by martijnlammers** [reply]
I happen to have written some shellscripts to automate the process of installation/setting up app development/config files. You still would have to change some of the configs as some things can't be assigned generically, like your DNS/san.
[you can find it here](https://github.com/mAlyanak/ccfsetup). By no means is it professional, but it might help you out. I'll  put a manual in the repo.
You can install CCF by running: 

> $ cd ~
> $ git clone https://github.com/martijnlammers/ccfsetup
> $ bash ./ccfsetup/ccf_install.sh

**Reply by ebo1981-cyber** [reply]
At first I am grateful to Mr. mAlyanak
Your guide is great, but during implementation I had a problem and I am unable to solve it ,when install npm
![image](https://user-images.githubusercontent.com/76199903/143531742-5947c485-99e2-4ba3-adb8-9645ef92f78f.png)
![image](https://user-images.githubusercontent.com/76199903/143542198-1aaed8ef-d5a3-489c-954c-bd72f22c8827.png)

**Reply by martijnlammers** [reply]
Sorry for the late response. I hope you managed to figure it out in the time I haven't responded. When I tested all of the scripts on my own virtual machine nothing seemed to be running into any errors. 

The reason this might have occurred is due to you perhaps using the shellscript more than just once, which would append multiple json bodies to thesame file. That would explain the sudden extra bracket. This is something I should've taken into consideration. My apologies.

You can resolve this by either removing all the shell script files and starting over (Cause you might have appended to different files as well) or by checking each file and removing the duplicates.

> }   <--- remove anything above this
> {
>     "private":true,...'


Since my reply, I've changed the original script, so it should now only append the json just one time, as it deletes any pre-existing file with thesame name before appending/creating package.json ensuring this error doesn't replicate itself anymore.
